src/i_midipipe.c

```diff
@@ -0,0 +1,481 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// DESCRIPTION:
+//     Client Interface to Midi Server
+//
+
+#if _WIN32
+
+#include <stdlib.h>
+#include <sys/stat.h>
+
+#define WIN32_LEAN_AND_MEAN
+#include <windows.h>
+
+#include "i_midipipe.h"
+
+#include "config.h"
+#include "i_sound.h"
+#include "i_timer.h"
+#include "m_misc.h"
+#include "net_packet.h"
+
+#if defined(_DEBUG)
+#define DEBUGOUT(s) puts(s)
+#else
+#define DEBUGOUT(s)
+#endif
+
+//=============================================================================
+//
+// Public Data
+//
+
+// True if the midi proces was initialized at least once and has not been
+// explicitly shut down.  This remains true if the server is momentarily
+// unreachable.
+boolean midi_server_initialized;
+
+// True if the current track is being handled via the MIDI server.
+boolean midi_server_registered;
+
+//=============================================================================
+//
+// Data
+//
+
+#define MIDIPIPE_MAX_WAIT 500 // Max amount of ms to wait for expected data.
+
+static HANDLE  midi_process_in_reader;  // Input stream for midi process.
+static HANDLE  midi_process_in_writer;
+static HANDLE  midi_process_out_reader; // Output stream for midi process.
+static HANDLE  midi_process_out_writer;
+
+//=============================================================================
+//
+// Private functions
+//
+
+//
+// FreePipes
+//
+// Free all pipes in use by this module.
+//
+static void FreePipes()
+{
+    if (midi_process_in_reader != NULL)
+    {
+        CloseHandle(midi_process_in_reader);
+        midi_process_in_reader = NULL;
+    }
+    if (midi_process_in_writer != NULL)
+    {
+        CloseHandle(midi_process_in_writer);
+        midi_process_in_writer = NULL;
+    }
+    if (midi_process_out_reader != NULL)
+    {
+        CloseHandle(midi_process_out_reader);
+        midi_process_in_reader = NULL;
+    }
+    if (midi_process_out_writer != NULL)
+    {
+        CloseHandle(midi_process_out_writer);
+        midi_process_out_writer = NULL;
+    }
+}
+
+//
+// UsingNativeMidi
+//
+// Enumerate all music decoders and return true if NATIVEMIDI is one of them.
+//
+// If this is the case, using the MIDI server is probably necessary.  If not,
+// we're likely using Timidity and thus don't need to start the server.
+//
+static boolean UsingNativeMidi()
+{
+    int decoders = Mix_GetNumMusicDecoders();
+
+    for (int i = 0;i < decoders;i++)
+    {
+        if (strcmp(Mix_GetMusicDecoder(i), "NATIVEMIDI") == 0)
+        {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+//
+// WritePipe
+//
+// Writes packet data to the subprocess' standard in.
+//
+static boolean WritePipe(net_packet_t *packet)
+{
+    DWORD bytes_written;
+    BOOL ok = WriteFile(midi_process_in_writer, packet->data, packet->len,
+        &bytes_written, NULL);
+
+    if (!ok)
```

Comment:
I wrote it that way because I didn't consider Win32 `BOOL` and Chocolate Doom `boolean` to be the same type.  However, I suppose when you peel back the curtain, they do turn out to be nearly the same thing - Win32 `BOOL` is an integer, and uses defines to designate `TRUE` and `FALSE` as 1 and 0 respectively, while Chocolate Doom `boolean` uses an enum, which should be an integer.  If it's close enough for you, it's close enough for me.

`bytes_written` is a required parameter to pass in under the circumstances that I am using it, and supplying a null value results in a crash on some versions of Windows.  According to MSDN, an incomplete write (`TRUE` return value with bytes written < bytes to write) only appears to be possible with a non-blocking pipes, and anonymous pipes block when the buffer is full.